block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -133,57 +135,100 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
   def insert(
       block: BlockMessage,
       invalid: Boolean
-  ): F[BlockDagRepresentation[F]] =
-    lock.withPermit(
+  ): F[BlockDagRepresentation[F]] = {
+    import cats.instances.list._
+    import coop.rchain.catscontrib.Catscontrib.ToBooleanF
+
+    val logAlreadyStored =
+      Log[F].warn(s"Block ${PrettyPrinter.buildString(block, short = true)} is already stored.")
+
+    def doInsert(blockMetadata: BlockMetadata): F[Unit] = {
+      val updateLatestMessages = {
+        val sendersNewLM         = (block.sender, block.blockHash)
+        val senderHasValidFormat = block.sender.size() == Validator.Length
+        val senderIsEmpty        = block.sender == ByteString.EMPTY // Genesis has empty sender
+        val logEmptySender =
+          Log[F].warn(s"Block ${Base16.encode(block.blockHash.toByteArray)} sender is empty")
+
+        /** For newly bonded validators there are several options. In general, we should be as lazy as possible.
+          * And probably should add LM only when newly bonded validator creates a block.
+          * But for now we set the block where validator has been bonded as its first Latest Message.
+          * This logic is subject to some research. **/
+        // If validator is bonded but does not have self-justification in block, this is newly bonded validator
+        val newlyBondedSet = bonds(block)
+          .map(_.validator)
+          .toSet
+          .diff(block.justifications.map(_.validator).toSet)
+
+        for {
+          // This filter is required to enable adding blocks backward from higher height to lower
+          newlyBondedUnseen <- newlyBondedSet.toList.filterA(
+                                lm => latestMessagesIndex.contains(lm).not
+                              )
+          newlyBondedLMs = newlyBondedUnseen.map(v => (v, block.blockHash)).toSet
+
+          // Do not create LM for empty sender, which is a now valid sender for genesis
+          lms <- if (senderIsEmpty || newlyBondedSet.contains(block.sender))
+                  newlyBondedLMs.pure[F]
+                // Update LM for sender
+                else {
+                  // LM should be updated only when block sequence number advances.
+                  val senderCurLMSeqNum = latestMessagesIndex
+                    .getUnsafe(block.sender)
+                    .flatMap(blockMetadataIndex.getUnsafe)
+                    .map(_.seqNum)
+                  val checkRegressiveSeqNum = senderCurLMSeqNum.map(block.seqNum < _)
+
+                  for {
+                    _ <- Sync[F]
+                          .raiseError[(ByteString, ByteString)](BlockSenderIsMalformed(block))
+                          // genesis has empty sender, so we have to put additional check
+                          .whenA(!senderHasValidFormat && !senderIsEmpty)
+                    _           <- logEmptySender.whenA(senderIsEmpty)
```

Comment:
`senderIsEmpty` is not used to decide if `sendersNewLM` will be added to `newlyBondedLMs` so it doesn't have to be part of the same logic.
Additional helper function can be useful because it can only calculate case when sender is not empty.